hwy/contrib/random/random.h

```diff
@@ -0,0 +1,219 @@
+    for (auto i = UINT64_C(0); i < thread_id; ++i) {
+      Jump();
+    }
+  }
+
+  constexpr std::uint64_t operator()() noexcept { return Next(); }
+
+  constexpr double Uniform() noexcept {
+    return static_cast<double>(Next() >> 11) * 0x1.0p-53;
+  }
+
+  constexpr std::array<std::uint64_t, 4> GetState() const {
+    return {m_state[0], m_state[1], m_state[2], m_state[3]};
+  }
+
+  constexpr void SetState(std::array<std::uint64_t, 4> state) noexcept {
+    m_state[0] = state[0];
+    m_state[1] = state[1];
+    m_state[2] = state[2];
+    m_state[3] = state[3];
+  }
+
+  static constexpr std::uint64_t StateSize() noexcept { return 4; }
+
+ private:
+  std::uint64_t m_state[4];
+
+  static constexpr std::uint64_t Rotl(const std::uint64_t x, int k) noexcept {
+    return (x << k) | (x >> (64 - k));
+  }
+
+  constexpr std::uint64_t Next() noexcept {
+    const std::uint64_t result = Rotl(m_state[0] + m_state[3], 23) + m_state[0];
+    const std::uint64_t t = m_state[1] << 17;
+
+    m_state[2] ^= m_state[0];
+    m_state[3] ^= m_state[1];
+    m_state[1] ^= m_state[2];
+    m_state[0] ^= m_state[3];
+
+    m_state[2] ^= t;
+
+    m_state[3] = Rotl(m_state[3], 45);
+
+    return result;
+  }
+
+ public:
+
+  /* This is the jump function for the generator. It is equivalent
+ to 2^128 calls to next(); it can be used to generate 2^128
+ non-overlapping subsequences for parallel computations. */
+  constexpr void Jump() noexcept {
+    constexpr std::uint64_t JUMP[] = {0x180ec6d33cfd0aba, 0xd5a61266f0c9392c,
+                                      0xa9582618e03fc9aa, 0x39abdc4529b1661c};
+    std::uint64_t s0 = 0;
+    std::uint64_t s1 = 0;
+    std::uint64_t s2 = 0;
+    std::uint64_t s3 = 0;
+    for (auto i : JUMP)
+      for (auto b = 0; b < 64; b++) {
+        if (i & std::uint64_t{1} << b) {
+          s0 ^= m_state[0];
+          s1 ^= m_state[1];
+          s2 ^= m_state[2];
+          s3 ^= m_state[3];
+        }
+        Next();
+      }
+
+    m_state[0] = s0;
+    m_state[1] = s1;
+    m_state[2] = s2;
+    m_state[3] = s3;
+  }
+
+};
+}  // namespace internal
+template <typename T>
+class HWY_CONTRIB_DLLEXPORT VectorXoshiro {
+ public:
+  explicit VectorXoshiro(const std::uint64_t seed) {
+    namespace hn = hwy::HWY_NAMESPACE;
+    internal::Xoshiro xoshiro{seed};
+    const auto lanes = hn::Lanes(hn::DFromV<T>());
+    auto stateArray = hwy::MakeUniqueAlignedArray<std::uint64_t>(StateSize());
+
+    for (auto i = 0UL; i < lanes; ++i) {
+      const auto state = xoshiro.GetState();
+      for (auto j = 0UL; j < internal::Xoshiro::StateSize(); ++j) {
+        const auto index = lanes * j + i;
+        stateArray[index] = state[j];
+      }
+      xoshiro.Jump();
+    }
+
+    for (auto i = 0UL; i < internal::Xoshiro::StateSize(); ++i) {
+      m_state[i] = hn::Load(hn::DFromV<T>(), &stateArray[i * lanes]);
+    }
+  }
+
+  constexpr auto operator()() { return Next(); }
+
+  static constexpr std::uint64_t StateSize() noexcept {
```

Comment:
This is going to be interesting. For SVE and RISC-V, the size cannot be constexpr. We can use hn::MaxLanes instead, but that upper bound is huge on RISC-V. Maybe better is to say there are always 8 streams, and this class loops over the 8 with step size Lanes()?

This will only use 512 bits of vectors but I think that's reasonable, and it would still work on the rare platforms with more. The key is to replace DFromV with hn::CappedTag<uint64_t, 8>.

Another approach to consider is that of https://github.com/google/highway/blob/master/hwy/contrib/sort/algo-inl.h#L277 : we can have static class functions and pass in the state as vectors, then the code is truly vector-length agnostic. I suspect that would work best.